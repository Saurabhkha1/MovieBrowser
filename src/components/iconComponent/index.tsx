import React from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';
import {IconProp} from '@fortawesome/fontawesome-svg-core';

type Props = {
  icon: IconProp;
  size?: number;
  color?: string;
};

export const IconComponent = ({icon, size, color}: Props) => {
  return <FontAwesomeIcon icon={icon} size={size} color={color} />;
};
