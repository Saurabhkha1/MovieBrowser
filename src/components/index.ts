export * from './iconComponent';
export * from './header';
export * from './movieComponent';
export * from './loader';
