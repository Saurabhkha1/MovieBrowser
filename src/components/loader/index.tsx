import {View, ActivityIndicator, StyleSheet} from 'react-native';
import React from 'react';
import {Colors} from '../../utils/colors';

type Props = {
  size?: 'full' | 'normal';
  color?: string;
};

export const Loader = ({size, color = Colors.black}: Props) => {
  return (
    <View style={size === 'full' ? styles.fullScreen : styles.normal}>
      <ActivityIndicator size={'large'} color={color} />
    </View>
  );
};

const styles = StyleSheet.create({
  fullScreen: {
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  normal: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
