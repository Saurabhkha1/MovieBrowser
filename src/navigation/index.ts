export * from './bottomTabs';
export * from './stack';
