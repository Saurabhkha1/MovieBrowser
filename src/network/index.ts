import axios from 'axios';
import {BASE_URL, BEARER_TOKEN} from '../utils/constants';

type Movie = {
  id: number;
  title: string;
  poster_path: string;
  release_date: string;
  vote_average: number;
  overview: string;
};

export const fetchMovies = async (
  endpoint: string,
  page: number = 1,
): Promise<Movie[]> => {
  const response = await axios.get(`${BASE_URL}/movie/${endpoint}`, {
    params: {language: 'en-US', page},
    headers: {
      Authorization: `Bearer ${BEARER_TOKEN}`,
      Accept: 'application/json',
    },
  });
  return response.data.results;
};
