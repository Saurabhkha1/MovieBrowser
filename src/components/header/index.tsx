import React, {useState, useRef, useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Animated,
  TextInput,
  Keyboard,
} from 'react-native';
import {moderateScale} from '../../utils/helper';
import {Colors} from '../../utils/colors';
import {IconComponent} from '../iconComponent';
import {IconProp} from '@fortawesome/fontawesome-svg-core';
import {Icons} from '../../utils/icons';

type Props = {
  title: string;
  icon: IconProp;
  onChangeText: (text: string) => void;
  searchValue: string;
};

export const Header = ({title, icon, onChangeText, searchValue}: Props) => {
  const [isInputVisible, setInputVisible] = useState(false);
  const slideAnim = useRef(new Animated.Value(-100)).current;
  const inputRef = useRef<TextInput>(null);

  useEffect(() => {
    Animated.timing(slideAnim, {
      toValue: isInputVisible ? 0 : -100,
      duration: 300,
      useNativeDriver: true,
    }).start(() =>
      isInputVisible ? inputRef?.current?.focus() : Keyboard.dismiss(),
    );
  }, [isInputVisible, slideAnim]);

  const toggleInputVisibility = () => {
    setInputVisible(!isInputVisible);
    onChangeText?.('');
  };

  return (
    <View style={styles.container}>
      <View style={styles.leftContainer}>
        <IconComponent
          icon={icon}
          size={moderateScale(25)}
          color={Colors.white}
        />
        <Text style={styles.heading}>{title}</Text>
      </View>
      <TouchableOpacity onPress={toggleInputVisibility}>
        <IconComponent
          icon={Icons.search}
          size={moderateScale(25)}
          color={Colors.white}
        />
      </TouchableOpacity>

      <Animated.View
        style={[styles.inputContainer, {transform: [{translateY: slideAnim}]}]}>
        <TextInput
          ref={inputRef}
          style={styles.input}
          value={searchValue}
          placeholder="Search for movies..."
          placeholderTextColor={Colors.grey}
          onChangeText={onChangeText}
        />
        <TouchableOpacity
          style={styles.closeInputButton}
          onPress={toggleInputVisibility}>
          <IconComponent icon={Icons.close} size={moderateScale(25)} />
        </TouchableOpacity>
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: moderateScale(10),
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: Colors.darkBlue,
    justifyContent: 'space-between',
    position: 'relative',
  },
  leftContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  heading: {
    color: Colors.white,
    fontSize: moderateScale(20),
    fontWeight: '500',
    marginLeft: moderateScale(10),
  },
  inputContainer: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: Colors.white,
    padding: moderateScale(5),
    paddingHorizontal: moderateScale(10),
    borderTopWidth: moderateScale(1),
    borderTopColor: Colors.grey,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  input: {
    height: moderateScale(40),
    borderColor: Colors.grey,
    borderWidth: moderateScale(1),
    borderRadius: moderateScale(5),
    paddingHorizontal: moderateScale(10),
    color: Colors.black,
    width: '90%',
  },
  closeInputButton: {
    width: '10%',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
