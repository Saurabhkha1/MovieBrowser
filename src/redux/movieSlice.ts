import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {fetchMovies} from '../network';

export type Movie = {
  id: number;
  title: string;
  poster_path: string;
  release_date: string;
  vote_average: number;
  overview: string;
};

export type MoviesState = {
  nowPlaying: Movie[];
  popular: Movie[];
  topRated: Movie[];
  upcoming: Movie[];
  status: 'idle' | 'loading' | 'succeeded' | 'failed';
  error: string | null;
  page: number;
  hasMore: boolean;
};

const initialState: MoviesState = {
  nowPlaying: [],
  popular: [],
  topRated: [],
  upcoming: [],
  status: 'idle',
  error: null,
  page: 1,
  hasMore: true,
};

export const fetchMoviesByCategory = createAsyncThunk(
  'movies/fetchMoviesByCategory',
  async ({
    category,
    page,
  }: {
    category: 'now_playing' | 'popular' | 'top_rated' | 'upcoming';
    page: number;
  }) => {
    const movies = await fetchMovies(category, page);
    return {category, movies, page};
  },
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    resetState: state => {
      state.nowPlaying = [];
      state.popular = [];
      state.topRated = [];
      state.upcoming = [];
      state.status = 'idle';
      state.error = null;
      state.page = 1;
      state.hasMore = true;
    },
  },
  extraReducers: builder => {
    builder
      .addCase(fetchMoviesByCategory.pending, state => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(
        fetchMoviesByCategory.fulfilled,
        (
          state,
          action: PayloadAction<{
            category: string;
            movies: Movie[];
            page: number;
          }>,
        ) => {
          state.status = 'succeeded';
          if (action.payload.movies.length === 0) {
            state.hasMore = false;
          }
          switch (action.payload.category) {
            case 'now_playing':
              if (action.payload.page === 1) {
                state.nowPlaying = action.payload.movies;
              } else {
                state.nowPlaying = [
                  ...state.nowPlaying,
                  ...action.payload.movies,
                ];
              }
              break;
            case 'popular':
              if (action.payload.page === 1) {
                state.popular = action.payload.movies;
              } else {
                state.popular = [...state.popular, ...action.payload.movies];
              }
              break;
            case 'top_rated':
              if (action.payload.page === 1) {
                state.topRated = action.payload.movies;
              } else {
                state.topRated = [...state.topRated, ...action.payload.movies];
              }
              break;
            case 'upcoming':
              if (action.payload.page === 1) {
                state.upcoming = action.payload.movies;
              } else {
                state.upcoming = [...state.upcoming, ...action.payload.movies];
              }
              break;
            default:
              break;
          }
          state.page = action.payload.page;
        },
      )
      .addCase(fetchMoviesByCategory.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message || 'Failed to fetch movies';
      });
  },
});

export const {resetState} = moviesSlice.actions;
export default moviesSlice.reducer;
