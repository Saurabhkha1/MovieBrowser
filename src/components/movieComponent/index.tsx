import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Image,
  Platform,
} from 'react-native';
import React from 'react';
import {deviceWidth, moderateScale} from '../../utils/helper';
import {Colors} from '../../utils/colors';
import {Movie} from '../../redux/movieSlice';
import {IMAGE_BASE_URL} from '../../utils/constants';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../navigation';

type Props = {
  movie: Movie;
};

export const MovieComponent = ({movie}: Props) => {
  const navigation =
    useNavigation<NativeStackNavigationProp<RootStackParamList>>();
  return (
    <TouchableOpacity
      style={styles.container}
      onPress={() => navigation.navigate('MovieDetails', {movie: movie})}>
      <Image
        source={{
          uri: IMAGE_BASE_URL + movie?.poster_path,
        }}
        style={styles.poster}
      />
      <View style={styles.details}>
        <Text style={styles.name} numberOfLines={2} ellipsizeMode="tail">
          {movie?.title}
        </Text>
        <Text style={styles.dateWithNoMargin}>Release Date:</Text>
        <Text style={styles.date}>{movie?.release_date}</Text>
        <Text style={styles.date}>
          Rating: {movie?.vote_average?.toFixed(1)}/10
        </Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: (deviceWidth - moderateScale(30)) / 2,
    height: Platform.OS === 'android' ? moderateScale(450) : moderateScale(400),
    borderRadius: moderateScale(10),
    backgroundColor: Colors.lightBlue,
    marginRight: moderateScale(10),
    marginBottom: moderateScale(10),
  },
  poster: {
    width: '100%',
    height: '70%',
    borderRadius: moderateScale(10),
  },
  details: {
    padding: moderateScale(5),
  },
  name: {
    color: Colors.black,
    fontSize: moderateScale(15),
    fontWeight: '500',
    marginBottom: moderateScale(10),
  },
  date: {
    color: Colors.black,
    fontSize: moderateScale(15),
    marginBottom: moderateScale(10),
  },
  dateWithNoMargin: {
    color: Colors.black,
    fontSize: moderateScale(15),
  },
});
