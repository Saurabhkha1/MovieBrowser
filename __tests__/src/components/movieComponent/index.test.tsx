import React from 'react';
import {create, act, ReactTestRenderer} from 'react-test-renderer';
import {Image, TouchableOpacity, Text} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {MovieComponent} from '../../../../src/components';
import {Movie} from '../../../../src/redux/movieSlice';

const mockMovie: Movie = {
  id: 1,
  title: 'Test Movie',
  poster_path: '/testpath.jpg',
  release_date: '2022-01-01',
  vote_average: 8.5,
  overview: '',
};

describe('MovieComponent', () => {
  it('renders correctly with given movie', () => {
    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<MovieComponent movie={mockMovie} />);
    });

    const tree = component?.toJSON();
    console.log(tree);
    expect(tree).toMatchSnapshot();

    const image = component?.root.findByType(Image);
    const titleText = component?.root
      .findAllByType(Text)
      .find(node => node.props.style && node.props.style.fontWeight === '500');
    const releaseDateText = component?.root.findByProps({
      children: 'Release Date:',
    });

    expect(image?.props.source.uri).toBe(
      `https://image.tmdb.org/t/p/w500${mockMovie.poster_path}`,
    );
    expect(titleText?.props.children).toBe(mockMovie.title);
    expect(releaseDateText).toBeTruthy();
  });

  it('navigates to MovieDetails on press', () => {
    const navigateMock = jest.fn();
    (useNavigation as jest.Mock).mockReturnValue({navigate: navigateMock});

    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<MovieComponent movie={mockMovie} />);
    });

    const touchable = component?.root.findByType(TouchableOpacity);
    act(() => {
      touchable?.props.onPress();
    });

    expect(navigateMock).toHaveBeenCalledWith('MovieDetails', {
      movie: mockMovie,
    });
  });

  it('renders correctly with different movie prop values', () => {
    const anotherMockMovie: Movie = {
      id: 2,
      title: 'Another Test Movie',
      poster_path: '/anotherpath.jpg',
      release_date: '2021-12-12',
      vote_average: 7.3,
      overview: '',
    };

    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<MovieComponent movie={anotherMockMovie} />);
    });

    const tree = component?.toJSON();
    console.log(tree); // Log the component tree
    expect(tree).toMatchSnapshot();

    const titleText = component?.root
      .findAllByType(Text)
      .find(node => node.props.style && node.props.style.fontWeight === '500');
    expect(titleText?.props.children).toBe(anotherMockMovie.title);
  });
});
