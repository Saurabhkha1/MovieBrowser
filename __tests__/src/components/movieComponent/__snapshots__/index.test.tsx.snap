// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieComponent renders correctly with different movie prop values 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderRadius": 10,
      "height": 400,
      "marginBottom": 10,
      "marginRight": 10,
      "opacity": 1,
      "width": NaN,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://image.tmdb.org/t/p/w500/anotherpath.jpg",
      }
    }
    style={
      {
        "borderRadius": 10,
        "height": "70%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "padding": 5,
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "fontWeight": "500",
          "marginBottom": 10,
        }
      }
    >
      Another Test Movie
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
        }
      }
    >
      Release Date:
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "marginBottom": 10,
        }
      }
    >
      2021-12-12
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "marginBottom": 10,
        }
      }
    >
      Rating: 
      7.3
      /10
    </Text>
  </View>
</View>
`;

exports[`MovieComponent renders correctly with given movie 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderRadius": 10,
      "height": 400,
      "marginBottom": 10,
      "marginRight": 10,
      "opacity": 1,
      "width": NaN,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://image.tmdb.org/t/p/w500/testpath.jpg",
      }
    }
    style={
      {
        "borderRadius": 10,
        "height": "70%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "padding": 5,
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "fontWeight": "500",
          "marginBottom": 10,
        }
      }
    >
      Test Movie
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
        }
      }
    >
      Release Date:
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "marginBottom": 10,
        }
      }
    >
      2022-01-01
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 15,
          "marginBottom": 10,
        }
      }
    >
      Rating: 
      8.5
      /10
    </Text>
  </View>
</View>
`;
